histomicsui/web_client/views/HierarchyWidget.js

```diff
@@ -1,13 +1,153 @@
+
 import { wrap } from '@girder/core/utilities/PluginUtils';
+import { restRequest } from '@girder/core/rest';
+import AccessWidget from '@girder/core/views/widgets/AccessWidget';
 import HierarchyWidget from '@girder/core/views/widgets/HierarchyWidget';
 import ItemCollection from '@girder/core/collections/ItemCollection';
+import UserCollection from '@girder/core/collections/UserCollection';
+import AnnotationCollection from '@girder/large_image_annotation/collections/AnnotationCollection';
 
 wrap(HierarchyWidget, 'initialize', function (initialize, settings) {
     settings = settings || {};
     if (settings.paginated === undefined) {
         settings.paginated = true;
     }
-    return initialize.call(this, settings);
+
+    initialize.call(this, settings);
+
+    if (this.parentModel.get('_modelType') === 'folder') {
+        fetchCollections(this, this.parentModel.id);
+    }
+    this.folderListView.on('g:folderClicked', () => {
+        fetchCollections(this, this.parentModel.id);
+        addAccessControl(this);
+    });
+});
+
+wrap(HierarchyWidget, 'render', function (render) {
+    render.call(this);
+
+    if (this.parentModel.get('_modelType') === 'folder' && this.recurseCollection) {
+        addAccessControl(this);
+    }
 });
 
+function fetchCollections(root, folderId) {
+    restRequest({
+        type: 'GET',
+        url: 'annotation/folder/' + folderId + '/present',
+        data: {
+            id: folderId,
+            recurse: true
+        }
+    }).done((resp) => {
+        if (resp[0]) {
+            root.users = new UserCollection();
+
+            root.recurseCollection = new AnnotationCollection([], {comparator: null});
+            root.recurseCollection.altUrl = 'annotation/folder/' + folderId;
+            root.recurseCollection.fetch({
+                id: folderId,
+                sort: 'created',
+                sortDir: -1,
+                recurse: true
+            }).done(() => {
+                root.recurseCollection.each((model) => {
+                    root.users.add({'_id': model.get('creatorId')});
+                });
+                $.when.apply($, root.users.map((model) => {
+                    return model.fetch();
+                })).always(() => {
+                    root.render();
+                });
+            });
+
+            root.collection = new AnnotationCollection([], {comparator: null});
+            root.collection.altUrl = 'annotation/folder/' + folderId;
+            root.collection.fetch({
+                id: folderId,
+                sort: 'created',
+                sortDir: -1,
+                recurse: false
+            }).done(() => {
+                root.collection.each((model) => {
+                    root.users.add({'_id': model.get('creatorId')});
+                });
+                $.when.apply($, root.users.map((model) => {
+                    return model.fetch();
+                })).always(() => {
+                    root.render();
```

Comment:
I'm seeing multiple render calls which makes the page flash quite a bit.  I'll run down what is going on.